packages/blocks/src/store/process-block-type.js

```diff
@@ -56,11 +57,14 @@ export const processBlockType =
 			selectors: {},
 			supports: {},
 			styles: [],
-			variations: [],
 			blockHooks: {},
 			save: () => null,
-			...select.getBootstrappedBlockType( name ),
+			...bootstrappedBlockType,
 			...blockSettings,
+			variations: [
+				...( bootstrappedBlockType?.variations || [] ),
+				...( blockSettings?.variations || [] ),
+			],
```

Comment:
I initially planned to use `deepmerge` here, but then I decided to handle this singular case for now.

Reply:
I think you should do some more sophisticated merging from the beginning. It doesn't need to be deep merge, but it should handle the following typical situation.

The server usually knows some subset of the block registration info. It can know as little as just the `name` and `render_callback`. The bootstrapped block info from server can look something like this:
```js
{
  name: 'core/group',
  title: 'Group',
  variations: [
    {
      name: 'group-row',
      title: 'Row'
    },
    {
      name: 'group-stack',
      title: 'Stack'
    }
  ]
}
```
Then when the client registers the block in JS, it has the _complete_ information. Including some fields that have JS functions as values, and can't be even expressed in PHP and JSON. The client-side block registration then does a shallow merge of the server and client values:
```js
...bootstrappedBlockType, // from server PHP
...blockSettings, // from client JS
```
There's no merging of fields, just overwriting server values with client ones.

The individual `variations` have the same field structure as the top-level block. Therefore, to combine server and client information, you should match the array elements by `name`, and merge them. Like:
```js
variations = bootstrappedVariations.map( b => {
  return {
    ...b,
    ...clientVariations.find( c => c.name === b.name ),
  }
} );
```
The real code should also add the items that are not in `bootstrappedVariations` but are only in `clientVariations`, but you get the idea.

Note that there is no deep merging, just matching the array elements and shallow-merging them together.

Reply:
Makes sense. Thanks, @jsnajdr!

Also, just to clarify, I was considering deep merging of whole setting objects.

```js
...deepmerge( bootstrappedBlockType, blockSettings )
```

Reply:
I think `deepmerge` is overkill, and base block registration doesn't do that either. You might even end up with some unexpected buggy results. It might not always be a good idea to merge `{a}` and `{b}` into `{a,b}` or `{a:[1]}` and `{a:[2]}` into `{a:[1,2]}`.

Deep merging should be very specifically targeted at specific properties. For example, with `variations`, we'll be doing one-level deep merge of two arrays keyed by `name`.

Reply:
> I think deepmerge is overkill

I thought so too, and dropped the idea 😄

We can add similar handlers as needed and eventually extract them into a custom merger method.

Reply:
Updated in [ad4c38d](https://github.com/WordPress/gutenberg/pull/60832/commits/ad4c38d4cbaca70053e1d5e14a62d788c1b3be40).